### PR TITLE
[IMP] account_edi, l10n_*: record limit for single job

### PR DESCRIPTION
**Pull request by lordkrandel**
If you:
- generate a lot (i.e. 10000) of records inside account.edi.documents model
- call the _cron_process_documents_web_services or the ir_cron_edi_network cron.

The current behavior is that only one single job with all records (i.e. 10000) are generated. This can take a lot of time or synchronization problem for cron process the queue.

We implement a document_size limit parameter for a single job so that the cron can enqueue items in buckets.

This is probably an issue that has arisen after we got the batching and the move_applicability.

Links: odoo/odoo#148849, https://www.odoo.com/web#model=project.task&id=3677341
issue-148849
opw-3677341

### CODE REVIEW

**Comment by smetl** [comment]
Imo, there is nothing to fix here. You are not supposed to reach this case since the cron will retrigger itself again and again until there is nothing left to process.

**Comment by lordkrandel** [comment]
@smetl 

We have a `job_count` to limit the number of `jobs`, but one job is free to process 100 invoices a day and if it takes too much time it gets killed - so we have no retrigger. Tomorrow it will have a double number of invoices to process, and it will fail again.

I don't feel comfortable when PRs that come from tickets or issues get dismissed, could you please suggest what could be wrong in the case, and how to fix eventually?

**Comment by smetl** [comment]
@lordkrandel The easiest way would be to limit the number of fetched documents. However, some EDI needs a custom grouping key. For them, we need to fetch all available documents.
For the same reason, your custom limit could break the grouping key, specially with a given number of 5 documents at the same time. Put at least 30, 40 or 50?
The main issue is for pe_edi I think since we have a cancel flow in 2 steps in which we need to send the same invoices for the two steps.

**Comment by lordkrandel** [comment]
> @lordkrandel The easiest way would be to limit the number of fetched documents. However, some EDI needs a custom grouping key.

Yes, that's why I avoided it

> For them, we need to fetch all available documents. For the same reason, your custom limit could break the grouping key, specially with a given number of 5 documents at the same time. Put at least 30, 40 or 50?  

As per the PR description:

>> _(document_count=5 is to be changed, it's to start the conversation)_

My original idea was 50, for now I'm just testing with less 'cause it's easier.

> The main issue is for pe_edi I think since we have a cancel flow in 2 steps in which we need to send the same invoices for the two steps.

Hmmm... So the same invoice is processed twice in the same job/batching key? O.o'' 
I think I have to investigate a bit how `pe_edi` works. Thanks for pointing it out 👍🏻

**Comment by lordkrandel** [comment]
@antoine162 The document limit PR

**Comment by antoine162** [comment]
@smetl I told @lordkrandel I can continue work on this PR so he can work on other stuff.

I understand the need in l10n_pe_edi to make sure that when we retrieve the cancellation ticket's status from SUNAT, this is taken into account as the 'definitive' status of the cancellation, even if we're no longer able to query the status of the ticket afterwards.

I think we can still achieve this while relaxing the constraint that all invoices with the same cancellation ticket are processed together at step 2. 

Indeed, at step 2 we can check whether any invoice with the same cancellation ticket was already cancelled (which means that step 2 already succeeded). In that case, we can just mark the invoice as cancelled, without querying the `getStatus` endpoint again. 

I wrote a PR for this in the enterprise branch -> https://github.com/odoo/enterprise/pull/54957

I tested it in 16.0. I performed cancellation step 1 in batch for 2 invoices, and then did step 2 separately for the two invoices. No problem.

Let me know your thoughts.

**Comment by antoine162** [comment]
Hello @smetl I thought about your comment that in Peru, we would want all invoices with a single cancellation request to be processed together. 

I guess this means we want different behaviours in Italy (limit invoices per batch) and Peru (make sure all invoices with the same grouping key are processed in the same batch). 

So, I've deleted the PR I had made for Peru, and I'm proposing that we add a key in `_get_move_applicability` that lets each EDI indicate whether a limit should be done for batching, and what that limit should be. 

Let me know your thoughts. :smile:

**Review by lordkrandel — COMMENTED** [review]
A little review

File: addons/account_edi/models/account_edi_document.py [hunk 1]
```diff
@@ -96,12 +100,19 @@ def _prepare_jobs(self):
                 else:
                     batching_key.append(move.id)
 
+                if move_applicability.get('batching_limit'):
+                    batching_key_without_limit = tuple(batching_key)
+                    batching_key.append(batching_limit_keys[batching_key_without_limit])
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_edi_document.py:105]: I think this doesn't really work: 

Suppose I have 9 documents, and the batch limit is 4.
For documents 1-4, I put them in a batch where `batching_key = [..., 0]`
Then, we have `len(batch['documents]) == limit`, so `current_batch_number = 1`
So, for document 5, I put it in a new batch where `batching_key = [..., 1]`.
But then, we again have `len(batch['documents']) == 1` so we get back to `current_batch_number = 0`
And then it doesn't work for documents 6-8. 

Basically, I think we sort of need to track how many batches we have so far for each grouping key, which is what my defaultdict does. 

(Of course, there could be other ways of doing this - first grouping without limit, and then splitting batches in a second step, but that would involve rewriting more of the method)

[the remark this answers is not in the document]
Comment (lordkrandel) [reply addons/account_edi/models/account_edi_document.py:105]: > But then, we again have len(batch['documents']) == 1,

Aaah, yeah, this is not a general count of the documents processed, but it's of course reset at every batch change.
Nevermind, my bad. 🤦🏻  Need sleep.

**Comment by lordkrandel** [comment]
@smetl  Poke  👉🏻

**Review by smetl — APPROVED** [review]
@robodoo r+
